ngraph/frontend/onnx_editor/src/edge_mapper.cpp

```diff
@@ -0,0 +1,218 @@
+//*****************************************************************************
+// Copyright 2017-2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include <onnx/onnx_pb.h>
+
+#include "ngraph/except.hpp"
+#include "onnx_editor/edge_mapper.hpp"
+
+using namespace ngraph;
+using namespace ngraph::onnx_editor;
+
+onnx_editor::EdgeMapper::EdgeMapper(const ONNX_NAMESPACE::GraphProto& graph_proto)
+{
+    int topological_index = 0;
+    m_node_inputs.resize(graph_proto.node().size());
+    m_node_outputs.resize(graph_proto.node().size());
+    for (const auto& node_proto : graph_proto.node())
+    {
+        for (const auto& out_name : node_proto.output())
+        {
+            // node output name is unique
+            m_node_name_to_index.emplace(out_name, topological_index);
+            m_node_outputs[topological_index].push_back(out_name);
+        }
+        for (const auto& in_name : node_proto.input())
+        {
+            m_node_inputs[topological_index].push_back(in_name);
+        }
+        if (!node_proto.name().empty())
+        {
+            // node name can identify node, but it can be ambiguous
+            m_node_name_to_index.emplace(node_proto.name(), topological_index);
+        }
+        ++topological_index;
+    }
+}
+
+std::vector<int> onnx_editor::EdgeMapper::find_node_indexes(const std::string& node_name,
+                                                            const std::string& output_name) const
+{
+    if (!output_name.empty())
+    {
+        const auto& index_iter = m_node_name_to_index.find(output_name);
+        if (index_iter != std::end(m_node_name_to_index))
+        {
+            return std::vector<int>{index_iter->second};
+        }
+    }
+    if (!node_name.empty())
+    {
+        std::vector<int> result;
+        auto matched_nodes_range = m_node_name_to_index.equal_range(node_name);
+        for (auto& index_iter = matched_nodes_range.first; index_iter != matched_nodes_range.second;
```

Comment:
I'm not sure but perhaps you can create a vector using both iterators returned from equal range(instead of the for-loop):
`const std::vector<int> result(matched_nodes_range.first, matched_nodes_range.second);`

Reply:
Contraction with std::vector seems to be not applicable in this case, but I've changed to `std::transform` using.